Add Set rules to GUI

viniciusbm opened the pull request:
GUI for #1410. This is currently a draft.

Limitations: 

- Device serial numbers and unit IDs are not validated. 
- Settings, keys and values are displayed even if the selected device does not support them.
- Some settings are not supported yet because I don't know their types and parameters.

pfps commented:
Excellent.  I'll take a look and provide feedback.

pfps commented:
For the device selection it might be good to grab the current device names and associated unitID or serial # and select using  the name (and ID if needed to disambiguate) but store the ID #.    The GUI then might not need the same-device switch at all.

Once a device has been selected it should be possible to get its settings and select from the GUI names for them and then the arguments for the selected setting.  This will require some work on settings_templates and settings.

Once the

pfps commented:
The code to get the device information would be like find_device in window.py except return the devices instead of checking for a matching device.

Once you have the device you can get the settings, which have both the internal name and the GUI name so you can display the GUI name but store the internal name.  Settings also have validators (_validator), which are of six kinds so you probably only need to implement them.  For many of the kinds (range, choice, map_choice, and you can get the possible values from information on the setting as follows:

Kind			key			value

toggle						True, False, Toggle
multiple_toggle		options, a seq of keys 	True, False, Toggle
choice			       		  	seq of values
map_choice		choices, a map from keys to a list of values
range				       	    	 int between min_value and max_value inclusive
multiple_range		....			 ....


keys and values are often NamedInt, which should be presented as the string and stored as the int, I think.


I don't understand how the gestures param setting works so leaving that undone seems reasonable.

viniciusbm commented:
Thank you for the feedback and the ideas. I'll make some changes and update the PR.

viniciusbm commented:
> This will require some work on settings_templates and settings.

For each setting I can test on my device, I created a subclass to replace the several pieces of code spread across *settings_templates.py*. I think this will be easier to maintain. Of course, if you don't like the idea of splitting into files, I can move the classes to a single file (*settings_templates.py* or a new file), which would still be better than the way it is.

pfps commented:
You seem to be making several changes to the settings template setup:

1/ make a subclass for each setting
2/ make the name, label, etc information be class variables
3/ move each setting to a separate file

I think the first is a good idea, and should be done for all the settings.  I was thinking about doing this anyway.   

I think the second is probably a good idea, but it requires a rewrite of the settings class and the settings setup code.
There is also the issue of what to do for settings that share name, label, and description.  I suppose that it is OK to have that information repeated.

I'm not keen on the third.


There is also the issue of where to put stuff like the mapping between bits and keys in the keyboard disable keys setting.  I think that this does not belong in the setting code, as it is really part of the HID++ specification and thus belongs either in hidpp20.py or special_keys.py.


So the code for settings should end up looking something like:

```
class DisableKeyboardKeys(BitFieldSetting):
    NAME = 'disable-keyboard-keys'
    LABEL = _('Disable keys')
    DESCRIPTION = _('Disable specific keyboard keys.')
    FEATURE = _FEATURE.KEYBOARD_DISABLE_KEYS
    RW = _FeatureRW(FEATURE, read_fnid=0x10, write_fnid=0x20)

    _labels = {k: (None, _('Disables the %s key.') % k) for k in _DKEY}

    def _callback(self, device):
        mask = device.feature_request(self.FEATURE)[0]
        options = [_DKEY[1 << i] for i in range(8) if mask & (1 << i)]
        return BitFieldValidator(options) if options else None


class RegisterHandDetection(Setting):
    NAME = 'hand-detection'
    LABEL = _('Hand Detection') 
    DESCRIPTION = _('Turn on illumination when the hands hover over the keyboard.')
    REGISTER = _REGISTER.keyboard_hand_detection
    RW = _RegisterRW(_R.keyboard_hand_detection)
    VALIDATOR = _BooleanV(true_value=b'\x00\x00\x00', false_value=b'\x00\x00\x30', mask=b'\x00\x00\xFF')


class LowresSmoothScroll(Setting):
    NAME = 'smooth-scroll'
    LABEL = _('Scroll Wheel Smooth Scrolling')
    DESCRIPTION = _('High-sensitivity mode for vertical scroll with the wheel.')
    FEATURE = _FEATURE.KEYBOARD_DISABLE_KEYS
    RW = _FeatureRW(_F.LOWRES_WHEEL)
    VALIDATOR = _BooleanV()
```

This should be done as a separate PR.  I should be able to take a crack at it today or tomorrow.

viniciusbm commented:
> I think the second is probably a good idea, but it requires a rewrite of the settings class and the settings setup code.
There is also the issue of what to do for settings that share name, label, and description. I suppose that it is OK to have that information repeated.

In that case, I think we could have an intermediate class that is extended by the settings that share those attributes. 

> I'm not keen on the third.

At first I thought each setting class would be larger, but after seeing your examples, in fact the classes will be short and should be in the same file. 

> I think that this does not belong in the setting code, as it is really part of the HID++ specification and thus belongs either in hidpp20.py or special_keys.py.

I agree. If they're part of the specification, it's possible that other settings will use them in the future.

viniciusbm commented:
> So the code for settings should end up looking something like

It looks much cleaner and easier to understand and maintain. 

We can also have unified names for the class attributes that represent
- the keys of map/multiple settings;
- the values of choice and map-choice settings;
- the minimum and maximum values of range settings. 

So the class attributes could have **all** of the possible or known keys and values, and the instance attributes (with slightly different names) would have just the ones that are available for the corresponding device. Thus, the rule editor could suggest and auto-complete using the available settings for the device when it's connected, but still validate the input using the full list if the device is unknown or not connected. What do you think? 

> This should be done as a separate PR. I should be able to take a crack at it today or tomorrow.

Great. I'll test your PR, and after it's merged, I'll rebase mine.

pfps commented:
I also like the general / specific argument checking.

pfps commented:
PR #1423 changes settings to classes.  

It should be possible to add the information you need to control what argument values are valid when the device is not available.  If you are unsure as to what the possible values are for some setting, let me know.

Also, the first two commits on this PR can be added independently.  Can you create a new PR for them?  Thanks.

viniciusbm commented:
Thanks, I'll try the new PR. 

> Also, the first two commits on this PR can be added independently. Can you create a new PR for them? Thanks.

Should it be based on master (a09af76) or on #1423 (951bc55)?

pfps commented:
It occurs to me that using code in the class to create the validator will cause problems because there is nothing that can be examined before the setting is created.  If, however, there is data in the class that is used to create the validator then this data can be examined.  I've already done this for the reader/writer.  Let me try it out for the validator.

pfps commented:
OK, PR #1423 has been updated.  Look there for information on how to get the information I think you need.

viniciusbm commented:
I'll see it now, thanks.

viniciusbm commented:
> keys and values are often NamedInt, which should be presented as the string and stored as the int, I think.

Yes, we have to store them as numbers, because some of the `NamedInt`s are created using translated names. Storing them as strings would make the rules unusable if the user changes the program language or if a translated is added after an update.

pfps commented:
NamedInts can also change for other reasons.  Some of the names of keys are not official Logitech names and sometimes the Logitech names are not good enough.  So there have been updates to these names.

viniciusbm commented:
Now we use the new setting classes. 

Still a draft. Device is not checked yet, and the other fields are validated using the full list.

pfps commented:
Looks good so far.

pfps commented:
Looking good. Here are a few issues.

Sensitivity (DPI) has too few choices, and the saved value looks unusual (maybe they come from the register version). I think the problem here is that the setting name is shared between two settings that are quite different. I think the best solution might be to break them apart by renaming the register one to something like 'dpi-old' 'Sensitivity (older mice)'. Users with devices that use this setting will need to redo it.

The choices for SetOS contains both generic and device-specific choices. I think that if there are device-specific choices only those should be available.

The range for Scroll Wheel Ratchet should be 0 to 50, I think. This should already be available in both the setting class and in settings.

I don't like having two widgets for the device. Would it be possible to have a widget that shows the current on-line devices plus a special value for originating device but also allows for serial/unit ID entry (8 hex digits)? When an on-line device is selected only valid choices for that device's settings are allowed.

viniciusbm commented:
Thanks for the comments. 

> I don't like having two widgets for the device. Would it be possible to have a widget that shows the current on-line devices plus a special value for originating device but also allows for serial/unit ID entry (8 hex digits)? 

Yes, I'll do that. 

> I think the problem here is that the setting name is shared between two settings that are quite different. I think the best solution might be to break them apart by renaming the register one to something like 'dpi-old' 'Sensitivity (older mice)'.

That's a good idea. When two settings with the same name have a different set of keys or values that isn't a subset, it isn't possible to have a correct list of keys/values if the numbers overlap. Should `DualPlatform` or `Multiplatform` also be renamed as well?

pfps commented:
The problem with requiring choices to be a NamedInts is that they are then sorted.  This doesn't work well for key reprogramming, where it is useful to have the normal value first.

pfps commented:
Let's leave Dual/Multi Platform alone for now.

viniciusbm commented:
> The problem with requiring choices to be a NamedInts is that they are then sorted. This doesn't work well for key reprogramming, where it is useful to have the normal value first.

I can't test that feature, but the last commit uses a subclass of NamedInts and should preserve the order (with the normal action first).

pfps commented:
Works for me, at least for the actual Key/Button Action setting on my MX Master 3.

pfps commented:
Things to be done:  
- split the two DPI settings
- add choices for the DPI feature-based setting based on the possible values from the MX Master 3 (by 50s from 200 to 4000 inclusive)
- single widget for device selection
- use actual setting keys and values when an on-line device is selected

Let me know if you want me to do the first two, as I can check the feature-based DPI setting.

viniciusbm commented:
> I can check the feature-based DPI setting.

Thanks. I made the two changes to the DPI settings, and the widget now shows the expected values. Do the bytes correspond exactly to the numbers or is there some transformation like in the old setting (`(x - 0x80) * 100`)?

pfps reviewed (changes requested):
Lower end should be 200.

pfps commented:
The device stuff looks better but when I switch between different set actions the device from the old action is copied to the new one.  
It would be nice if the warning was toned down when an 8-digit hex ID is entered manually.

viniciusbm commented:
> It would be nice if the warning was toned down when an 8-digit hex ID is entered manually.

Should the warning icon be hidden if the string matches `[A-F0-9]{8}` even if current devices don't have that id?

pfps commented:
Hmm.   Ideally a different icon, but not nothing.  If it is hard to do a different icon then don't bother.

pfps commented:
That looks great!

pfps commented:
I notice that the Device field can be too small for the Device information.  Is there a good way to have to widgets expand?

pfps commented:
That looks better.

I guess the next thing is to get the actual settings and permissable keys/values for on-line devices and use them instead of the "universal" ones.

pfps commented:
When I switch between Key/Button Diversion and Key/Button Actions the value stays the same even though it is not suitable for the new setting.

pfps commented:
I'm seeing errors on startup (with a bunch of junky rules).
```
idefix /home/local/SoftwareDownloads/Solaar> bin/solaar -d
Traceback (most recent call last):
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/__init__.py", line 180, in _status_changed
    diversion_rules.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1882, in update_devices
    _diversion_dialog.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 677, in update_devices
    rc.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1745, in update_devices
    device_value = self.collect_value()[0]
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1828, in collect_value
    if kind in self.MULTIPLE or kind is None and len(self.component.args) > 3:
AttributeError: 'NoneType' object has no attribute 'args'
Traceback (most recent call last):
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/__init__.py", line 180, in _status_changed
    diversion_rules.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1882, in update_devices
    _diversion_dialog.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 677, in update_devices
    rc.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1745, in update_devices
    device_value = self.collect_value()[0]
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1828, in collect_value
    if kind in self.MULTIPLE or kind is None and len(self.component.args) > 3:
AttributeError: 'NoneType' object has no attribute 'args'
Traceback (most recent call last):
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/__init__.py", line 180, in _status_changed
    diversion_rules.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1882, in update_devices
    _diversion_dialog.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 677, in update_devices
    rc.update_devices()
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1745, in update_devices
    device_value = self.collect_value()[0]
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1828, in collect_value
    if kind in self.MULTIPLE or kind is None and len(self.component.args) > 3:
AttributeError: 'NoneType' object has no attribute 'args'
```

pfps commented:
If I don't set a key for a Key/Button Actions setting when I switch back to the setting I see:
```
Traceback (most recent call last):
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 290, in _selection_changed
    self.ui[type(component)].show(component, wrapped.editable)
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/ui/diversion_rules.py", line 1814, in show
    self.key_field.get_child().set_text(key)
TypeError: Argument 1 does not allow None as a value
```

pfps commented:
Looks like the last two bugs are fixed.

pfps commented:
Here is another problem.  I have two devices where only the unitID is reported and the rule editor thinks they are the same device.
Sometimes serial numbers are known for devices but sometimes they aren't.  Sometimes unit IDs are known for devices but sometimes they aren't.  The two are supposed to be the same but sometimes they aren't.  
So Solaar has to keep track of both serial number and unit ID and accept either as an identifier for the device.  Solaar also stores the serial number as a text string and uses '?' as the default.
It looks as if the rule editor just uses serial number and thinks '?' is a valid number.
What I can do is add an identifier function that hides this detail, but it still means that in some cases the indentifier will be different.  I can also try to use unitID more often, so that the cases where different identifiers are reported become very rare.

viniciusbm commented:
Currently, the editor stores the serial number if available and the unitId otherwise. The same value is shown by the widget in parentheses. Both are accepted if typed by the user. 

> It looks as if the rule editor just uses serial number and thinks '?' is a valid number.

Thanks, I didn't know that. I'll consider '?' as unavailable, and that should be enough to fix the bug for devices that have at least one of those identifiers.

pfps commented:
OK for now.

viniciusbm commented:
> Sometimes serial numbers are known for devices but sometimes they aren't. Sometimes unit IDs are known for devices but sometimes they aren't.

If a device doesn't have either identifier, it's not supported by Set action according to the documentation. In that case, should the editor ignore it?

viniciusbm commented:
I've noticed that `_status_changed` is called many times per event (e.g. if I power on my keyboard, it is called four times).

To avoid executing the same code several times unnecessarily, I think the rule editor UI could cache the list of devices and their relevant metadata in memory. Thus, when `_status_changed` is called, the rule editor would do nothing unless there's a new device or a device that hasn't been seen online before has just been powered on. To do that, we'd have to assume that a device doesn't change its set of settings, keys, possible values, etc. Is this a safe assumption?

pfps commented:
Parts of Solaar are quite reactive.  When a device is turned on it can go through several states as far as Solaar is concerned, thus the multiple calls change its status when it would be better to wait.  Some of these calls even cause the Solaar window to update and this might be done too often.  But I don't think that much GTK work is done because the GUI caches the widgets it uses.   

Most information about most settings will not change.  Similarly the permissable settings will not change.   A device can only go from no settings known (initial status) to only a few settings known (never been online) to all settings known (has been online).   What can change is the current value of a setting (which the rule editor probably doesn't care about).   What can also change, for one setting only, is the values that Solaar displays in the GUI.  The setting is the Change Host setting where the values are a combination of the slot number (1-3) and information about the computer that the device connects to when using that slot.  Solaar doesn't update these values as it goes along, which is a bit of a bug.

Anyway, I think the rules editor can depend on the possible settings, the possible values, the possible keys, and the possible values for a key all not changing and ignore the case above especially because the internal possible values don't change, just what Solaar displays.

pfps commented:
This looks very good.  Just one thing that isn't working right and a few things that might work better and finally some commentary about how the universe of choices in a setting class relates to the actual choices in the setting.


A few choice settings (e.g., Mouse Gestures) have an extra value, usually something like "Off" with value 0.  It appears that this extra value is not allowed.   This extra value is choices_extra in the setting class.  


If I type in the identifier for an existing device in upper case hex then no icon is shown in the device widget, and the rule the the main pane shows the device name.  Can the widget also show the device name?

If the identifier is in mixed case then it is not recognized as being the same as in upper case.  Can this be changed to work like upper case, i.e., it is recognized as an actual device?


Most choice widgets allow for free text entry.  That's useful for selecting from many choices.  But when you do type in one of the choices that isn't recognized as being a valid entry.  Can this be changed so that if the entered string is equal (case insensitive) to precisely one of the choices then the entered string is replaced by the choice?


Most choice widgets have an empty position.  But the boolean widget doesn't.  Should the boolean widget allow for an invalid empty value?  Or is this not easy because the boolean widget is implemented differently?



It appears that there is the assumption that all valid choices for a setting are elements of the choices universe for the setting class.   There are a few settings where this assumption isn't true.  It may be that the assumption is useful (e.g., for settings with many choices - and there is a new setting coming along where the size of the choice universe is about 8000), though, and so would be good to keep where it is true.  A suggestion would be to have a flag in the setting class indicating whether all choices belong to the choice universe and behave differently if the flag is false.  The current settings where the assumption isn't true are AdjustableDpi, ChangeHost, and Multiplatform/Dualplatform.  It may be that my earlier message related to choices wasn't clear that this could be the case.

All three of these settings are different and may need different solutions.   I think that the best solution may be to just pretend that the universe for AdjustableDpi is the real universe and only have the flag for the two others, which have small sets of values.  If the flag is set then when there is a known device the possible values are taken directly from the possible values for the device without requiring them to be in the universe.

viniciusbm commented:
> This looks very good.

Thanks for the review and the suggestions. 

> I don't understand how the gestures param setting works so leaving that undone seems reasonable.

Implemented in 82428e757dfc4675c09999a8ee3f8a75e089f84a and cdc97a98cd4f0add95007a39554ca3cd049c7322.

> A few choice settings (e.g., Mouse Gestures) have an extra value, usually something like "Off" with value 0. It appears that this extra value is not allowed. This extra value is choices_extra in the setting class.

Fixed by af558ead1a51387525fed1b671acb88559e3bea7. 

> If I type in the identifier for an existing device in upper case hex then no icon is shown in the device widget, and the rule the the main pane shows the device name. Can the widget also show the device name?

Implemented in e9bedfc17baf46ea5e806869fb0e58df59268776.

> If the identifier is in mixed case then it is not recognized as being the same as in upper case. Can this be changed to work like upper case, i.e., it is recognized as an actual device?

Implemented in 66d9e95db9ed78ee2ed602834c66f656e9c75476.

> Most choice widgets allow for free text entry. That's useful for selecting from many choices. But when you do type in one of the choices that isn't recognized as being a valid entry.

Fixed by 19606a67d01d8dcb47b11ab2f659891f6052e685. 

> Can this be changed so that if the entered string is equal (case insensitive) to precisely one of the choices then the entered string is replaced by the choice?

Implemented in de6b8280cb9f1e514b2f935a294ccd0bb01850b0. 

> Most choice widgets have an empty position. But the boolean widget doesn't. Should the boolean widget allow for an invalid empty value?

Changed in a3f1aa4a0ba589ce2dfde4030a01b2f97cd92447.

> It appears that there is the assumption that all valid choices for a setting are elements of the choices universe for the setting class. There are a few settings where this assumption isn't true.

If I understand correctly, this is something I had already implemented, but it was disabled. I enabled it in 45c64b1db2be031e08a3a64ac4bcff6770df9d5f. Now, if I temporarily comment out some keys in *special_keys.py* for the setting Disable Keys, the unknown values are still displayed with their fallback names when the device is connected. Are the other cases also covered by this change?

pfps commented:
Excellent work.   

There is one remaining glitch out of the changes.   When a setting for a device has fewer possibilities than the universe then only the device-specific possibilities are shown.  A setting that works like this is dpi.  
But when the device has other possibilities these are shown along with the setting class possibilities.  They should replace these possibilities even when the value is the same, I think.  An example of this is Set OS, where the setting class possibilities come from two setting classes with different user-visible strings but intersecting internal numbers.  If there is no device to determine what options to show it is good to show all the possibilities.  When there is a device, though, only its possibilities should be shown, even if they share an internal number with a possibilities from the setting classes.   
So Set OS for my Craft keyboard should only show three choices (0, Windows), (1, MacOS), (2, IOS) even though multiplatform has (0, OS 1).  Similarly Change Host should not show Host X because this is not a good choice for my Craft even though it has possible values that have the same internal number.

pfps commented:
One other change to consider.
Set rules are shown with external strings (e.g., Back Button instead of 83) in some places but with internal values in others (e.g., reprogrammable-keys instead of Key/Button Actions).  This should probably be the external string everywhere.

viniciusbm commented:
> Excellent work.

Thank you. 

> But when the device has other possibilities these are shown along with the setting class possibilities. They should replace these possibilities even when the value is the same, I think. 

It should be fixed in 1cfa9b9, but since I can't test it, I'll probably need some help to debug if it doesn't work. 

> Set rules are shown with external strings (e.g., Back Button instead of 83) in some places but with internal values in others (e.g., reprogrammable-keys instead of Key/Button Actions). This should probably be the external string everywhere.

Changed in bb662cf.

pfps commented:
Something broke in these last two commits.  The value for a setting that is a choice doesn't show up in the top part of the rule editor, but is saved in the rules.yaml file.

viniciusbm commented:
> Something broke in these last two commits. The value for a setting that is a choice doesn't show up in the top part of the rule editor, but is saved in the rules.yaml file.

Fixed in fc6ca54.

pfps commented:
| It should be fixed in 1cfa9b9, but since I can't test it, I'll probably need some help to debug if it doesn't work


This doesn't work because NamedInts __contains__ is permissive, i.e., a NamedInt is in a NamedInts if the number or the string matches.
I added a new method to NamedInts in common.py
```
    def has_element(self, value):
        return self[value] == value
```
and updated show_only to end as
```
        self.set_all_values(values, (lambda v: only is None or ( only.has_element(v) if isinstance(only, NamedInts) else v in only)))
```
feel free to modify as appropriate.

There are two related issues remaining.  

The matcher for values triggers on the internal value, not just the user strings.  So typing '1' into the Change Host setting results in NamedInt(1,'Host 2') being selected.  This happens if Host 2 isn't one of the choices shown.  It looks as if the matching should take into account which choices are visible and needs to be done as a string, not an integer, for NamedInts.

The top part of the rule editor displays a generic string (e.g., 'Host 2') instead of a device-specific string (e.g., '2:idefix').  The universe contains NamedInt(1,'Host 2') but the device-specific values include NamedInt(1, '2:idefix').  Device-specific values should be preferred.  The problem here may be that the device-specific values have not been set up yet.

viniciusbm commented:
> I added a new method to NamedInts in common.py

Since `__eq__` in `NamedInt` is also permissive, I can't see why that method would be different than just using checking `__contains__` in `NamedInts`. 

```
>>> A = NamedInts(x=1, y=2)
>>> ('x' in A, A.has_element('x'))
(True, True)
>>> ('z' in A, A.has_element('z'))
(False, False)
>>> (1 in A, A.has_element(1))
(True, True)
>>> (6 in A, A.has_element(6))
(False, False)
>>> ('1' in A, A.has_element('1'))
(False, False)
```

Am I missing something? 


> The matcher for values triggers on the internal value, not just the user strings.

I explicitly implemented it that way, which I thought wouldn't be a problem because I didn't know some keys/values would have numbers in their names that do not correspond to their numerical values. I'll change it so that the numerical values are not accepted when typed.

pfps commented:
What is happening is that __contains__ checks for intergerness and sringness only but __eq__ first checks for NamedIntness first and in this case requires both parts to be equal.  The right solution is to fix __contains__ but I worry about breaking something that depends on how __contains__ works so I prefer getting rules to work and only afterwards fixing __contains__

viniciusbm commented:
I think the field works now. I'll fix the text displayed on the top part.

viniciusbm commented:
I see there are still some issues affecting both the text field and the top part when a device is connected. I'll update the code soon.

viniciusbm commented:
It seems to be working now.

pfps commented:
I'm still seeing the following behaviour for Change Host:

The settings class choices are NamedInt(N, 'Host N+1) for n from 0 to 2.
The device settings choices are things like NamedInt(N, 'N+1: system name') for n from 0 to 2.

The editing widget shows the choices from the device settings.
The top part of the pane shows the Host N+1 values instead of the device settings.

Other than that, things look good.

viniciusbm commented:
I added the same logic of the editing widget. Now, if the setting instance has a non-empty `NamedInts` as `choices`, then it is used to generate the text even if the value is unknown (a fallback name is generated), and the class attribute is not used.

pfps commented:
Things look good now.  I'll do some more testing in a while, but this might be done!

pfps reviewed (commented):
Looks good, but I still need to do some more testing.

viniciusbm commented:
> Things look good now. I'll do some more testing in a while, but this might be done!

Thanks!

pfps commented:
I tested on several devices and everything I tried looked good.  I'll try to get some other testers.

pfps commented:
@barolo @sk8geek @mschilli87  This PR is ready for testing.  
It mostly involves the UI for editing Set rules.   To test out, click on "Rule Editor" in the Solaar main window.  Then click on "User-defined rules" and select "Insert new rule".  Then click on "[empty]"  and select Insert Here -> Action -> Set.  You should see a rule action that sets a setting on a device, initially Originating device '' None.  Go to the bottom of the window and use the Device field to select one of your devices.  Then go to the Setting field and select one of the settings there - only settings for that device should be shown.  Try several of the settings and see what Values or Keys and Values are available - they should only be those that make sense for your device, i.e., the ones shown in the Solaar main window.  Click on "Save changes" and then open ~/.config/solaar/rules.yaml.  You should see a rule there with a Set: action, the ID number of your device, the internal name for the setting, and one or two other arguments depending on whether the setting has a key.  The arguments may be internal numbers for the setting so don't worry too much about them.
Finally quit Solaar, start it up again, start the rule editor, and check that the rule you created shows up and makes sense.   Thanks.

To clone and use Solar from its GitHub repository
```
git clone https://github.com/pwr-Solaar/Solaar.git
cd Solaar
```
Run Solaar as bin/solaar from this directory.


To run PR #1422, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/1422/head:pull_1422
git checkout pull_1422
```
To download a new version of the pull request, fetch it and then set your pull branch
to the new fetch, as in:
```
git checkout pull_1422
git fetch origin pull/1422/head
git reset --hard FETCH_HEAD

pfps commented:
The structure of this PR is a bit messed up, with changes to some underlying stuff mixed in with the GUI changes.  If you don't mind, I'm going to try to clean it up a bit.

pfps reviewed (commented):
This commit should probably be pulled out into a separate PR.

viniciusbm commented:
> could you pull out the translation commit and submit it as a separate PR?

Most of the strings I translated were created by the changes in this PR (because I wanted to be sure that the displayed texts are not the internal names), so it wouldn't make sense to have the translation first. I can remove it and submit it later as a separate PR after this one is merged. What do you think? 

It might be a good idea, especially because there seems to be something that needs to be checked. The mechanism that generates the .po files seems to ignore some `pgettext` calls, so I had to manually add the corresponding lines to the .po file. The use of that function was necessary because the translation of "key" depends on the context.

pfps commented:
Yes please pull the translation out and put in in later as a separate PR.  Thanks.

pfps commented:
Do you know where the problem might be with not picking up pgettext?

pfps commented:
The four commits that fix `solaar config` appear to be separable as well.  I cherry-picked them into a separate branch and then rebased on that branch with no problem.  If it's fine with you I can merge these four commits from PR #1434.  Your authorship of the commits is retained.

pfps commented:
The other thing that might be worthwhile pulling out is the changes to NamedInts.  This is going to be a bit harder as one of the changes is not in a commit by itself.  Don't feel that this is something that needs to be done.

viniciusbm commented:
> Do you know where the problem might be with not picking up pgettext?

No, I haven't looked into it yet. 

> The four commits that fix solaar config appear to be separable as well. I cherry-picked them into a separate branch and then rebased on that branch with no problem. If it's fine with you I can merge these four commits from PR #1434. Your authorship of the commits is retained.

Yes, that's fine. Of course I'll have to rebase, but I would have to do it anyway now that #1423 was merged and GitHub already says that there are conflicts (because your commit from #1423 was included here).

viniciusbm commented:
I've just created a full backup of the current state, and I'll try to make the suggested changes.

pfps commented:
I did some more testing on this so I'm happy with it in its current form.
@viniciusbm If you would like to take a larger role in Solaar please let me know at pfpschneider@gmail.com
